livekit-plugins/livekit-plugins-google/livekit/plugins/google/llm.py

```
@@ -0,0 +1,298 @@
+        fnc_ctx: Optional[llm.FunctionContext] = None,
+        temperature: Optional[float] = None,
+        n: Optional[int] = 1,
+        **kwargs,
+    ) -> "LLMStream":
+        system_instruction = _extract_system_instruction(chat_ctx)
+
+        tools = _build_tools(fnc_ctx) if fnc_ctx and fnc_ctx.ai_functions else []
+
+        generation_config = GenerationConfig(
+            temperature=temperature,
+            candidate_count=n,
+            max_output_tokens=1024,
+        )
+
+        model = GenerativeModel(
+            model_name=self._model_name,
+            system_instruction=system_instruction,
+            **self._kwargs,
+        )
+
+        contents = _build_vertex_context(chat_ctx, id(self))
+
+        cmp = model.generate_content_async(
+            contents=contents,
+            generation_config=generation_config,
+            tools=tools or None,
+            stream=True,
+        )
+
+        return LLMStream(
+            llm=self,
+            vai_stream=cmp,
+            chat_ctx=chat_ctx,
+            fnc_ctx=fnc_ctx,
+        )
+
+
+def _extract_system_instruction(chat_ctx: llm.ChatContext):
+    system_messages = [msg.content for msg in chat_ctx.messages if msg.role == "system"]
+    if system_messages:
+        return system_messages
+    return None
+
+
+def _build_tools(fnc_ctx: llm.FunctionContext) -> List[Tool]:
+    tools = [
+        Tool(
+            function_declarations=[
+                FunctionDeclaration(
+                    name=fnc_info.name,
+                    description=fnc_info.description,
+                    parameters=_build_parameters(fnc_info.arguments),
+                )
+            ]
+        )
+        for fnc_info in fnc_ctx.ai_functions.values()
+    ]
+    return tools
+
+
+def _build_parameters(arguments: Dict[str, llm.FunctionArgInfo]) -> Dict[str, Any]:
+    properties = {
+        arg_name: {
+            "type": JSON_SCHEMA_TYPE_MAP.get(arg_info.type, "string"),
```

I don't think we should default to string?

---

Let's raise an error instead if the type isn't supported